src/coordinates/uniform_coordinates.hpp

```diff
@@ -0,0 +1,326 @@
+    }
+    area_[0] = dx_[1] * dx_[2];
+    area_[1] = dx_[0] * dx_[2];
+    area_[2] = dx_[0] * dx_[1];
+    cell_volume_ = dx_[0] * dx_[1] * dx_[2];
+  }
+  UniformCoordinates(const UniformCoordinates<System> &src, int coarsen)
+      : istart_(src.GetStartIndex()) {
+    dx_ = src.Dx_();
+    xmin_ = src.GetXmin();
+    xmin_[0] += istart_[0] * dx_[0] * (1 - coarsen);
+    xmin_[1] += istart_[1] * dx_[1] * (1 - coarsen);
+    xmin_[2] += istart_[2] * dx_[2] * (1 - coarsen);
+    dx_[0] *= coarsen;
+    dx_[1] *= (istart_[1] > 0 ? coarsen : 1);
+    dx_[2] *= (istart_[2] > 0 ? coarsen : 1);
+    area_[0] = dx_[1] * dx_[2];
+    area_[1] = dx_[0] * dx_[2];
+    area_[2] = dx_[0] * dx_[1];
+    cell_volume_ = dx_[0] * dx_[1] * dx_[2];
+  }
+  //----------------------------------------
+  // Dxc: Distance between cell centers
+  //----------------------------------------
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dx() const {
+    static_assert(dir > 0 && dir < 4);
+    return dx_[dir - 1];
+  }
+
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dxc(const int idx) const {
+    static_assert(dir > 0 && dir < 4);
+    return dx_[dir - 1];
+  }
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dxc(const int k, const int j, const int i) const {
+    static_assert(dir > 0 && dir < 4);
+    if constexpr (dir == X1DIR) return static_cast<const System *>(this)->template Dxc<dir>(i);
+    else if constexpr (dir == X2DIR) return static_cast<const System *>(this)->template Dxc<dir>(j);
+    else if constexpr (dir == X3DIR) return static_cast<const System *>(this)->template Dxc<dir>(k);
+    PARTHENON_FAIL("Unknown dir.");
+    return 0.0;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION Real Dxc(const int dir, const int k, const int j, const int i) const {
+    assert(dir > 0 && dir < 4);
+    if (dir == X1DIR) return static_cast<const System *>(this)->template Dxc<X1DIR>(k, j, i);
+    else if (dir == X2DIR) return static_cast<const System *>(this)->template Dxc<X1DIR>(k, j, i);
+    else if (dir == X3DIR) return static_cast<const System *>(this)->template Dxc<X1DIR>(k, j, i);
+    PARTHENON_FAIL("Unknown dir.");
+    return 0.0;
+  }
+
+  //----------------------------------------
+  // Dxf: Distance between cell faces
+  //----------------------------------------
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dxf(const int idx) {
+    static_assert(dir > 0 && dir < 4);
+    return dx_[dir - 1];
+  }
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dxf(const int k, const int j, const int i) const {
+    static_assert(dir > 0 && dir < 4);
+    return dx_[dir - 1];
+  }
+
+  //----------------------------------------
+  // Xc: Positions at cell centroids
+  //----------------------------------------
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real Xc(const int idx) const {
+    static_assert(dir > 0 && dir < 4);
+    return xmin_[dir - 1] + (idx + 0.5) * dx_[dir - 1];
+  }
```

Comment:
looks like you're shadowing these in derived classes. Should they be marked `overload` or something? Or is that not necessary due to CRTP?